src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs

```diff
@@ -43,7 +43,17 @@ public RegexBoyerMoore(string pattern, bool caseInsensitive, bool rightToLeft, C
             // We're doing this for your own protection. (Really, for speed.)
             Debug.Assert(pattern.Length != 0, "RegexBoyerMoore called with an empty string. This is bad for perf");
             Debug.Assert(pattern.Length <= MaxLimit, "RegexBoyerMoore can take a long time for large patterns");
-            Debug.Assert(!caseInsensitive || pattern.ToLower(culture) == pattern, "RegexBoyerMoore called with a pattern which is not lowercased with caseInsensitive true.");
+#if DEBUG
+            if (caseInsensitive)
+            {
+                foreach (char c in pattern)
+                {
+                    // We expect each individual character to have been lower-cased. We don't validate the whole
+                    // string at once because the rest of the library doesn't currently recognize/support surrogate pairs.
+                    Debug.Assert(c == culture.TextInfo.ToLower(c), "Pattern wasn't lowercased with provided culture");
```

Comment:
>Assert [](start = 26, length = 6)

This check is fine I want to point we can do that too by calling CharUnicodeInfo.GetUnicodeCategory and check if it is upper case category. I think your check is better here as this code is in debug build only and no perf concern with it.